src/shell/commands/head.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use std::fs::File;
+use std::io::Read;
+
+use anyhow::bail;
+use anyhow::Result;
+use futures::future::LocalBoxFuture;
+use tokio_util::sync::CancellationToken;
+
+use crate::ExecuteResult;
+use crate::ShellCommand;
+use crate::ShellCommandContext;
+use crate::ShellPipeWriter;
+
+use super::args::parse_arg_kinds;
+use super::args::ArgKind;
+
+pub struct HeadCommand;
+
+impl ShellCommand for HeadCommand {
+  fn execute(
+    &self,
+    context: ShellCommandContext,
+  ) -> LocalBoxFuture<'static, ExecuteResult> {
+    let mut stderr = context.stderr.clone();
+    let result = match execute_head(context) {
+      Ok(result) => result,
+      Err(err) => {
+        let _ = stderr.write_line(&format!("head: {err}"));
+        ExecuteResult::from_exit_code(1)
+      }
+    };
+    Box::pin(futures::future::ready(result))
+  }
+}
+
+fn execute_head(mut context: ShellCommandContext) -> Result<ExecuteResult> {
+  fn copy_lines<F: FnMut(&mut [u8]) -> Result<usize>>(
+    writer: &mut ShellPipeWriter,
+    max_lines: u64,
+    cancellation_token: &CancellationToken,
+    mut read: F,
+  ) -> Result<ExecuteResult> {
+    let mut lines = 0;
+    let mut buf = vec![0; 512];
+    while lines < max_lines {
+      if cancellation_token.is_cancelled() {
+        return Ok(ExecuteResult::for_cancellation());
+      }
+      let size = read(&mut buf)?;
+      if size == 0 {
+        break;
+      } else if let Some(line) = buf.split(|&b| b == b'\n').next() {
```

Comment:
What happens to the rest of the lines in the buffer? I'm not sure how the tests work so I must be reading this wrong.